File: Prácticas/deepspace/src/deepspace/SpecificDamage.java
package deepspace;

import java.util.ArrayList;
import java.util.stream.Collectors;

/**
 * SpecificDamage is a damage composed by 
 *  - the number of shields to lose
 *  - a list of weapon types to lose
 *
 * @author Miguel Ángel Fernández Gutiérrez, Sergio Quijano Rey
 */
class SpecificDamage extends Damage{
    /**
     * Array of types of weapons that will be lost.
     */
    ArrayList<WeaponType> weapons;

    // -------------------------------------------------------------------------
    // Constructors
    // -------------------------------------------------------------------------
    
    /**
     * Class initializer.
     * @param _weapons array of types of weapons that will be lost
     * @param _nShields number of shields that will be lost
     */
    public SpecificDamage(ArrayList<WeaponType> _weapons, int _nShields){
        super(_nShields);
        if( _weapons != null )
            weapons = new ArrayList<>(_weapons);    // Secure copy
        else
            weapons = new ArrayList<>(0);           // Security check
    }

    /**
     * Copy constructor.
     * @param other instance which is going to be copied
     */
    public SpecificDamage(SpecificDamage other){
        this(other.getWeapons(), other.getNShields());

    }

    // -------------------------------------------------------------------------
    // Getters
    // -------------------------------------------------------------------------
    
    /**
     * Copy getter.
     * @return a copy of the instance
     */
    @Override
    public SpecificDamage copy(){
        return new SpecificDamage(this);
    }
    
    /**
     * Getter for weapons
     * @return a copy (secure return) of weapons
     **/
    public ArrayList<WeaponType> getWeapons(){
        if ( weapons != null )
            return new ArrayList<WeaponType>(weapons);
        else {
            System.out.println("Returning null weapons at Damage.getWeapons()");
            System.out.println("Returning instead an empty ArrayList<WeaponType>");
            return new ArrayList<WeaponType>(0);
        }
    }

    /**
     * Checks if the damage has no effect
     * @return true, if the damage has no effect
     *         false, if the damage has an effect
     **/
    @Override
    public boolean hasNoEffect(){
        if ( weapons != null )
            return weapons.size() == 0 && getNShields() == 0;
        else
            return getNShields() == 0;
    }

    
    /**
     * Get the string representation of the object
     * @return the string representation
     **/
    @Override
    public String toString(){
        return  "Damage(" +
                "weapons = " + weapons +
                ", nShields = " + getNShields() +
                ")";
    }

    /**
     * Gets the UI version of the object
     *
     * @return the UI version of the object
     * */
    @Override
    public SpecificDamageToUI getUIversion(){
        return new SpecificDamageToUI(this);
    }
   
    // -------------------------------------------------------------------------
    // Setters
    // -------------------------------------------------------------------------
    
    /**
     * Adjusts the damage to a given weapons and shields
     * The number of shields is decreased depending of the given list of shields
     * The weapons which are not the same type as given as a parameter are removed
     *
     * @param w, the list of weapons to adjust
     * @param s, the list of shields to adjust
     * @return a new SpecificDamage adjusted as specified above
     **/
    @Override
    public SpecificDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s){
        // New number of shields is get, as much shields lost as s indicates
        int new_shields = Math.min(s.size(), getNShields());

            // New weapon_type list is get
            ArrayList<WeaponType> new_weapon_types = new ArrayList<WeaponType>(0);

            for(WeaponType wt : weapons){
                if(arrayContainsType(w, wt) != -1){ // The weapon type is in the w array
                    new_weapon_types.add(wt);
                }
            }

        return new SpecificDamage(new_weapon_types, new_shields);
    }

    /**
     * A given weapon is discarded from the Damage
     *
     * @param w, the weapon to discard
     * */
    @Override
    public void discardWeapon(Weapon w){
        weapons = new ArrayList<WeaponType>(weapons.stream().filter(wtype -> wtype != w.getType()).collect(Collectors.toList()));

    }
    
    // AUXILIAR
    //==========================================================================
    
    /**
     * Searchs the first weapon with the same type as given type through parameter
     * 
     * @param w, the weapon array where to make the search
     * @param t, the weapon type to search
     * @return  the position, if the weapon is found
     *          -1, if the weapon is not found
     * */
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t){
        int position = -1;

        for(int i = 0; i < w.size() && position == -1; i++){
            if(w.get(i).getType() == t){
                position = i;
            }
        }

        return position;
    }
}

File: Prácticas/Deepspace/lib/SpecificDamage.rb
#encoding:utf-8

require_relative 'Damage'
require_relative 'SpecificDamageToUI'


module Deepspace

# Class to represent specific damage
# @see Deepspace::Damage
#
# @author Miguel Ángel Fernández Gutiérrez, Sergio Quijano Rey
class SpecificDamage < Damage

    public_class_method :new

    # Class initializer
	# @param _weapons [Array<WeaponType>] array of types of weapons that will be lost
	# @param _nShields [Integer] number of shields that will be lost
    def initialize(_weapons, _nShields)
        super(_nShields)

        # @!attribute [Array<WeaponType>] array of types of weapons that will be lost
        @weapons  = []
    end

    attr_reader :weapons

    # Checks whether the damage is affecting or not
    # @return [Boolean] true, if damage has no effect; false, otherwise
    # --Overriden
	def hasNoEffect
		return (super && @weapons.length == 0)
    end

    # Searches the first element of WeaponType array to match a given type
	# @param w [Array<Weapon>] the array of weapon types where we search
	# @param t [WeaponType] the type we're looking for
	# @return [Integer] position, if the element is found;
	#                   -1, if no element is found
	def arrayContainsType(w, t)
		i = 0
		w.each do |weapon_aux|
			if weapon_aux.type == t
				return i
			else
				i += 1
			end
		end

		# No element found
		return -1
	end
    
    # Creates a copy of current objet where weapons and shields which are not
	# included in arrays given as parameters are discarded. That's to say, we
	# shrink the Damage to the parameters
	# @param w [Array<Weapon>] weapons to fit
	# @param s [Array<ShieldBooster>] shields to fit
	# @return [SpecificDamage] a copy of the object adjusted as explained above
    # --Overriden
    def adjust(w, s)
        weapons_copy = @weapons.clone

		new_weapons = w.map do |weapon|
			weapons_copy.delete_at(weapons_copy.index(weapon.type) || weapons_copy.length)
		end

		new_weapons.compact!
        
        return self.class.new(new_weapons, super(s))
    end

    # Copy getter
    # @return [SpecificDamage] a copy of the current instance
    # --Overriden
    def copy
        return self.class.new(weapons, nShields)
    end

    # Removes a given type of weapon.
	# If a list of weapon types is not available (object is Numeric-constructed
	# instead of Specific-constructed), the number of weapons decreases by 1
    # @param w [Weapon] the weapon whose type wants to be removed
    # --Overriden
    def discardWeapon(w)
        if @weapons.length != 0
            position = @weapons.index(w.type)
            if position != nil
                @weapons.delete_at(position)
            end
        end
    end

    # String representation of the object
    # @return [String] string representation
    def to_s
        message = "[Specific Damage] -> Shields: #{@nShields}, Weapon types: " + getWeaponInfo
        return message
    end

    # To UI
    def getUIversion
        return SpecificDamageToUI.new(self)
    end
    
end # class NumericDamage

end # module Deepspace